dv/uvm/core_ibex/collect_results.py

```diff
@@ -4,78 +4,217 @@
+           f'{len(failing_tests)} FAILED'
 
-    raise RuntimeError('Strange first line ({!r})'.format(first_line))
+def gen_test_run_result_text(test_run_result: TestRunResult) -> str:
+    '''Generate a string describing a TestRunResult including the failure
+    message if the test did not pass'''
+    test_name_idx = f'{test_run_result.name}.{test_run_result.seed}'
+    test_underline = '-' * len(test_name_idx)
 
+    info_text = f'{test_name_idx}\n{test_underline}\n'
 
-def dump_log(path: str, dest: TextIO) -> None:
-    print('\nLog at {}:'.format(path), file=dest)
-    with open(path, 'r') as fd:
-        for line in fd:
-            dest.write('> ' + line)
+    if (test_run_result.binary):
+        info_text += f'Test binary: {test_run_result.binary}\n'
 
+    if (test_run_result.uvm_log):
+        info_text += f'UVM log: {test_run_result.uvm_log}\n'
 
-def box_comment(line: str) -> str:
-    hr = '#' * 80
-    return hr + '\n# ' + line + '\n' + hr
+    if (test_run_result.rtl_trace):
+        info_text += f'RTL trace: {test_run_result.rtl_trace}\n'
+
+    if (test_run_result.iss_trace):
+        info_text += f'ISS trace: {test_run_result.iss_trace}\n'
+
+    if (test_run_result.comparison_log):
+        info_text += f'Comparison log: {test_run_result.comparison_log}\n'
+
+    if (test_run_result.passed):
+        info_text += f'[PASSED]\n'
+    else:
+        info_text += f'{test_run_result.failure_message}\n'
+
+    return info_text
+
+def output_results_text(passing_tests: List[TestRunResult], failing_tests:
+        List[TestRunResult], dest: TextIO):
+    '''Write results in text form to dest'''
+
+    if failing_tests:
+        dest.write(box_comment('Details of failing tests'))
+        dest.write('\n')
+        for trr in failing_tests:
+            dest.write(gen_test_run_result_text(trr))
+            dest.write('\n')
+
+    if passing_tests:
+        dest.write(box_comment('Details of passing tests'))
+        dest.write('\n')
+        for trr in passing_tests:
+            dest.write(gen_test_run_result_text(trr))
+            dest.write('\n')
+
+    dest.write('\n')
+    dest.write(gen_summary_line(passing_tests, failing_tests))
+    dest.write('\n')
+
+def output_run_results_junit_xml(passing_tests: List[TestRunResult], failing_tests:
+        List[TestRunResult], junit_dest: TextIO, junit_merged_dest: TextIO):
+    '''Write results to JUnit XML
+
+    Two versions are produced a normal version and a merged version. In the
+    normal version there is a test suite per unique test name with a different
+    test case per seed run. In the merged version there is a single test case
+    under the test suite with information for the individual runs merged
+    together. This is to aid use of the Azure Pipelines JUnit dashboard, which
+    doesn't neatly handle the test suite/test case hierarchy
+    '''
+
+    all_tests = passing_tests + failing_tests
 
+    test_suite_cases = {}
+    merged_test_suite_info = {}
```

Comment:
Since these are keyed by the same thing, I'd suggest generating a single dict here with tuples as values. Then the loop below gets a bit nicer: I think the whole lot can be replaced by something like
```python3
    by_name = {}
    for trr in all_tests:
        merged, unmerged = \
            by_name.setdefault(trr.name, ([], {'stdout': '', 'failures': ''}))

        result_text = gen_test_run_result_text(trr)

        # Create a test case for the TestRunResult. stdout holds the text
        # describing the run. Add the same text to failures if the test failed.
        test_case = junit_xml.TestCase(f'{trr.name}.{trr.seed}')
        test_case.stdout = result_text

        merged['stdout'] += result_text

        if not trr.passed:
            test_case.add_failure_info(output=trr.failure_message)
            merged['failures'] += result_text

        unmerged.append(test_case)
```

Following uses look like this:
```python3
    test_suites = [junit_xml.TestSuite(name, unmerged) for
                   (name, (merged, unmerged)) in by_name.items()]
```